ClosedTypeHierarchyDiagnosticSuppressor.Tests/SwitchExpressionSuppressorTests.cs

```diff
@@ -107,6 +107,26 @@ public static int DoSwitch(Root root)
 
         return EnsureSuppressed(code, NullableContextOptions.Enable, allowRecords: true);
     }
+    
+    [Test]
+    public Task When_record_with_destructure_And_only_base_type_is_matched_Then_do_not_suppress()
```

Comment:
Both of the tests are the wrong way around. The test should define a hierarchy where `Leaf1` has a `object v` and then test that a match agains `Leaf1(string v)` is not exhaustive.

The fact that the current wrong tests succeed is due to a missing feature implementation in the `IsDeclarationPatternNonRestrictive` method. The method currently expects `containingSubpatternSyntax.NameColon` to be present, whereas with deconstruction it is not.

Please fix the tests and add the implementation accordingly.

The positive tests (that assert suppression) happen to succeed correctly because they use `var`. Had you written them with `object v`, they would have failed as well and revealed the missing implementation.

Reply:
Since there was no reaction from the PR creator and I do not want to have this hanging around, I have picked this up myself.
It seems things are not that simple after all.

1. The deconstruct pattern has nothing to do with records at all, but is about types which can be deconstructed, i. e. those with a blessed `Deconstruct` method.
2. There can be multiple Deconstruct methods for the same type.
3. Deconstruct methods for a type can be declared as extension methods.

Reply:
Sorry about that. I completely forgot about this PR (I haven't used C# much recently).